03-rnaseq/gene-id-annotation_rnaseq_01_ensembl.Rmd

```diff
@@ -0,0 +1,380 @@
+  dplyr::arrange(desc(entrez_id_count))
+
+# Let's look at the first 6 rows of our `multi_mapped` object
+head(multi_mapped_df)
+```
+
+Looks like we have some cases where 15 Entrez IDs mapped to a single Ensembl ID.
+We have a total of 235 out of 31882 Ensembl IDs with multiple mappings to Entrez IDs.
+With that rate, we can filter out the multi-mapped IDs if we want to keep 1:1 mappings for our downstream analyses by supplying the `"filter"` option to the `multiVals` argument of `mapIds()`.
+If you want to see an example of a `multiVals = "filter"` strategy see the [microarray gene identifier annotation example notebook](https://github.com/AlexsLemonade/refinebio-examples/tree/master/02-microarray/gene-id-annotation_microarray_01_ensembl.html).
+
+### Store all mapped values
+
+We are not sure which Entrez ID is relevant as yet, but we could store all of the mapped information for future use.
+In the next chunk, we show how we can collapse all the Entrez IDs into one column where we store them all for each unique Ensembl ID.
+
+```{r}
+final_mapped_df <- df %>%
+  # Collapse the mapped values in `mapped_list` into one column named `entrez_ids`
+  dplyr::mutate(entrez_ids = sapply(mapped_list, paste, collapse = ";")) %>%
+  # Select the `entrez_ids` column first, then everything else in the data frame
+  dplyr::select(entrez_ids, dplyr::everything()) %>%
+  # Now arrange by the new `entrez_ids` column
+  dplyr::arrange(entrez_ids)
+```
+
+Let's take a look at our new collapsed `entrez_ids` column in the `final_mapped_df` object.
+
+```{r}
+final_mapped_df %>%
+  # Filter `final_mapped_df` to include only the rows where `entrez_ids` values include the ";" character -- we know that these are the rows with multiple mapped values
+  dplyr::filter(stringr::str_detect(entrez_ids, ";")) %>%
+  # We only need a preview here
+  head()
+```
+
+You may have a list of Entrez IDs you are interested in, in which case, the above format may work best for you.
+In a different study, you may want the oldest Entrez ID (which is probably first), in which case, you can create a column that stores just the first mapped Entrez ID that comes up for each Ensembl ID.
+We will do this in the next section by re-running the `mapIds()` function with `multiVals = "first"`.
+
+### Map Ensembl IDs to gene symbols -- keeping only the first mapped value
+
+Let's re-run `mapIds()`, this time using `multiVals = "first"` and add this information to our `final_mapped_df` object.
+
+```{r}
+final_mapped_df <- data.frame(
+  "first_mapped_entrez_id" = mapIds(
+    org.Dr.eg.db, # Replace with annotation package for the organism relevant to your data
+    keys = df$Gene,
+    column = "ENTREZID", # Replace with the type of gene identifiers you would like to map to
+    keytype = "ENSEMBL", # Replace with the type of gene identifiers in your data
+    multiVals = "first" # This will keep only the first mapped value for each Ensembl ID
+  )
```

Comment:
```suggestion
  ), final_mapped_df
```

Then no `inner_join` is needed. (I think this should be fine). Unless we find it too confusing for users to follow? (II personally find it easier to follow than a join.

Reply:
I think this would result in a duplicated column if you leave the `rownames_to_column` step... but I like joins. They are a good thing to learn and a good habit because I have been burned too many times by functions that don't preserve order (`mapIds` seems to, but I still don't trust it).